lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp

```diff
@@ -34,6 +34,49 @@ Value getWaveN(ConversionPatternRewriter &rewriter, Location loc, Value wave,
 
 namespace SharedToDotOperandMFMA {
 
+/**
+ * @brief swizzling tensor element indexes according pattern encoded in
+ * SharedEncodingAttr
+ *
+ * @param rewriter
+ * @param loc
+ * @param row row of target tensor element related to the start of smemObj
+ * @param col col of target tensor element related to the start of smemObj
+ * @param smemObj shared memory object, contains info about tensor in LDS
+ * @param attr layout attribute, contains swizzling info
+ * @return swizzled row, col indexes in tensor notation
+ */
+std::pair<mlir::Value, mlir::Value>
+swizzleIndexes(ConversionPatternRewriter &rewriter, Location loc, Value row,
+               Value col, SharedMemoryObject smemObj, SharedEncodingAttr attr) {
+  (void)smemObj; // unused in current pattern
+  bool transposed = (attr.getOrder()[0] != 1);
```

Comment:
This is only true for blockA.
For blockB, the dim is k x n, i.e. dim 0 is `k` and dim 1 is `n`. So column-major means order[0] == 1.

Reply:
I do not understand what is the problem here.

Could you clarify why you mentioned blockA/blockB here in the first place?

Swizzling do not know anything about A/B operands/blocks and following dot operations.

Reply:
The `order` of the sharedEncodingAttr tells us which dim is the leading/contiguous dim when data is in shared memory. Here we want to check whether it is transposed. 
The convention in triton is that
- matrix A has dim M by K, where K is the leading dim --> `order = {1,0}`
- matrix B has dim K by N, where N is the leading dim --> `order = {0,1`

So whoever wants to query information about dims, it must distinguish between A and B.
For example,
1. [Here](https://github.com/openai/triton/blob/62aef58c176bd8db7672f32fc4ff41dac12a04bb/include/triton/Dialect/TritonGPU/IR/TritonGPUAttrDefs.td#L94) it tries to obtain the dim for `nonKDim` of A or B
2. [Here](https://github.com/openai/triton/blob/62aef58c176bd8db7672f32fc4ff41dac12a04bb/lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMMAv2.cpp#L572) it sets `kOrder` according to isA

Reply:
> The order of the sharedEncodingAttr tells us which dim is the leading/contiguous dim when data is in shared memory. Here we want to check whether it is transposed.

This is correct

> The convention in triton is that
> - matrix A has dim M by K, where K is the leading dim --> order = {1,0}
> - matrix B has dim K by N, where N is the leading dim --> order = {0,1

This is not the case for triton. Both A and B operands could be transposed.
You can verify it by looking at ttgir generated by [test_dot](https://github.com/ROCmSoftwarePlatform/triton/blob/7eb30ff7aa53b13a7646decd6daa9ac10e45c280/python/test/unit/language/test_core_amd.py#L1237) test in test_core_amd.py and [upstream version](https://github.com/openai/triton/blob/62aef58c176bd8db7672f32fc4ff41dac12a04bb/python/test/unit/language/test_core.py#L2003) in test_core.py

Pay attention to `col_a` and `col_b` attributes.

> So whoever wants to query information about dims, it must distinguish between A and B.
For example,
> 1. [Here](https://github.com/openai/triton/blob/62aef58c176bd8db7672f32fc4ff41dac12a04bb/include/triton/Dialect/TritonGPU/IR/TritonGPUAttrDefs.td#L94) it tries to obtain the dim for nonKDim of A or B
> 2. [Here](https://github.com/openai/triton/blob/62aef58c176bd8db7672f32fc4ff41dac12a04bb/lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMMAv2.cpp#L572) it sets kOrder according to isA

1. I struggle to understand what this variable is mean to be.
IMHO [this code](https://github.com/openai/triton/blob/62aef58c176bd8db7672f32fc4ff41dac12a04bb/include/triton/Dialect/TritonGPU/IR/TritonGPUAttrDefs.td#L114)  contradicts description of the variable
2. kOrder defines which axis of abstract tensor is K, later this code uses `order` to find actual memory mapping: https://github.com/openai/triton/blob/62aef58c176bd8db7672f32fc4ff41dac12a04bb/lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMMAv2.cpp#L405

Actually, i am not sure what information you mean, here is my understanding, what information about tensor and it's location in memory we have:
We have following information, which lives in different places:
1. Pointer to beginning of the memory buffer, strides along axies and offsets of beginning of current tensor to the beginning of underlying memory object (this is used to get slices of tensors without actually copying data) This info is stored in [SharedmemmoryObject](https://github.com/ROCmSoftwarePlatform/triton/blob/ac1e551e8a51ad5969325bdb5a0e75010dff70aa/lib/Conversion/TritonGPUToLLVM/Utility.h#L205)
2. Swizzling info describes how this SharedMemoryObject elements are swizzled, it is `vec`, `maxPhase`, `perPhase`. This info is stored in `SharedEncodingAttr`
3. Tensor shape and it's element type, this info is stored in tensor itself.
If tensor is not sliced, we can compute `SharedMemoryObject` strides from tensor shape and vice versa, but this is not true if tensor is just a part of some bigger object.